### PR TITLE
8339488: Extended NPE message doesn't handle CONSTANT_Dynamic

### PR DESCRIPTION
**Pull request by olivergillespie**
Extended NPE messages require walking the bytecode in the method where it was created, and simulating the effect on the stack. `ldc` for dynamic constants is not handled, which can lead to assertion failures in debug builds, or incorrect NPE messages in release builds (possibly worse with the right conditions?).

Simply handle dynamic constants by figuring out their return type so we know how they will affect the stack.

The new test fails without the fix in both release and debug builds. In release, it fails because the NPE message is wrong: `Exception in thread "main" java.lang.NullPointerException: Cannot invoke "Object.notify()" because "`, and in debug builds it hits this assert:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
# Internal Error (<redacted>/src/hotspot/share/interpreter/bytecodeUtils.cpp:660), pid=12677, tid=12678
# assert(false) failed: Unexpected tag
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339488](https://bugs.openjdk.org/browse/JDK-8339488): Extended NPE message doesn't handle CONSTANT_Dynamic (**Bug** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) Review applies to [9382a708](https://git.openjdk.org/jdk/pull/20889/files/9382a708b3196fedb2465b11129e0a9963d41305)
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**) Review applies to [1cf82008](https://git.openjdk.org/jdk/pull/20889/files/1cf820088582b032d0735cc1816afd12f4896305)
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**) Review applies to [9382a708](https://git.openjdk.org/jdk/pull/20889/files/9382a708b3196fedb2465b11129e0a9963d41305)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20889/head:pull/20889` \
`$ git checkout pull/20889`

Update a local copy of the PR: \
`$ git checkout pull/20889` \
`$ git pull https://git.openjdk.org/jdk.git pull/20889/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20889`

View PR using the GUI difftool: \
`$ git pr show -t 20889`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20889.diff">https://git.openjdk.org/jdk/pull/20889.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20889#issuecomment-2333852965)

### CODE REVIEW

**Comment by liach** [comment]
A few questions:
1. Can the reproduction case be as simple as:
```
ldc Dynamic(BootstrapMethod(ConstantBootstraps::nullConstant) NameAndType(ConstantDescs.DEFAULT_NAME, "Ljava/lang/Object;"))
invokevirtual Methodref(Class("java/lang/Object") NameAndType("notify" "()V"))
```
2. Is there a preference of ASM/Java generator vs hardcoded jasm? Think it's better to use one file for such a small test.

**Comment by olivergillespie** [comment]
> Can the reproduction case be as simple as:
...

Probably, I just don't know how to write that in jasm format. If you know how, I can change it.

> Is there a preference of ASM/Java generator vs hardcoded jasm? Think it's better to use one file for such a small test.

I am following the style of tests nearby. Can you share an example of your idea?

**Comment by olivergillespie** [comment]
Thanks, that helps. Using asm doesn't look that much simpler to me, and considering the other tests in the same folder are using jasm, I think we might as well stick with it.

**Review by simonis — APPROVED** [review]
The fix looks right. Thanks for finding an easy reproducer.

I'm fine with the jasm test.

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @requires vm.flagless
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib
+ * @compile CondyExtendedNullPointer.jasm
+ * @run driver CondyExtendedNullPointerTest
+ */
+
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.compiler.InMemoryJavaCompiler;
```
Comment (lmesnik) [inline test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:37]: Seems InMemoryJavaCompiler  is not used anymore.

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8339488
+ * @summary Test extended NullPointerException message in method with CONSTANT_Dynamic.
+ * @requires vm.flagless
```
Comment (lmesnik) [inline test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:28]: Is there are any reasons to ignore all additional vm flags? Usually the test should accept any flags and be flagless only if there are known reasons for this. Even if it doesn't looks to be useful to run in different modes. Usually to stress comiplers with different bycodes.
You don't need to run test in an additional configurations.

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java (hunk 2 above)
Comment (olivergillespie) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:28]: I copied from CondyLDCTest - I figured the authors must have thought about something I didn't.

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java (hunk 1 above)
Comment (olivergillespie) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:37]: True. I just copied from CondyLDCTest which I guess also has it unused. Will remove.

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java (hunk 2 above)
Comment (lmesnik) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:28]: I quickly looked through test history and didn't find any reason for this. Could you please change
'createLimitedTestJavaProcessBuilder'
to 
'createTestJavaProcessBuilder'
and remove 
line 28
@requires vm.flagless.

I'll file a bug to fix CondyLDCTest also.

**Review by lmesnik — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java (hunk 2 above)
Comment (lmesnik) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:28]: BTW, I filed PR and you might see that one of tests start crash. So it is really useful to accept vm flags.

**Review by coleenp — APPROVED** [review]
VM code looks good.  I'm fine with the test in jasm.

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java (hunk 2 above)
Comment (coleenp) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:28]: CondyLDCTest test crashes with VM flags?

Comment (lmesnik) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:28]: Sorry, I forgot to add link. Test runtime.condy.TestEscapeThroughInvokeWithCondy fails. See PR 
https://github.com/openjdk/jdk/pull/20894

**Review by lmesnik — APPROVED** [review]

**Review by liach — APPROVED** [review]
Don't be worried about my previous remarks.

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java (hunk 2 above)
Comment (olivergillespie) [reply test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java]: Thanks for looking. Updated as you suggested, @lmesnik, please take a quick look when you can.

**Review by lmesnik — APPROVED** [review]

File: test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java [hunk 3]
```diff
@@ -37,7 +36,7 @@
 
 public class CondyExtendedNullPointerTest {
     public static void main(String args[]) throws Throwable {
-        ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder("-Xverify:all",
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder("-Xverify:all",
                                                                              "CondyExtendedNullPointer");
```
Comment (ExE-Boss) [inline test/hotspot/jtreg/runtime/condy/CondyExtendedNullPointerTest.java:40]:
```suggestion
        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder("-Xverify:all",
                                                                      "CondyExtendedNullPointer");
```

**Pull request closed without merging**